lib/string.gi

```diff
@@ -642,23 +642,21 @@ InstallGlobalFunction("RemoveCharacters", REMOVE_CHARACTERS);
 ##
 #F  EvalString( <expr> ) . . . . . . . . . . . . evaluate a string expression
 ##
-_EVALSTRINGTMP := 0;
-InstallGlobalFunction("EvalString", function( s )
-  local a, f, res;
-  a := "_EVALSTRINGTMP:=";
-  Append(a, s);
-  Add(a, ';');
-  Unbind(_EVALSTRINGTMP);
-  f := InputTextString(a);
-  Read(f);
-  if not IsBound(_EVALSTRINGTMP) then
-    Error("Could not evaluate string.\n");
-  fi;
-  res := _EVALSTRINGTMP;
-  Unbind(_EVALSTRINGTMP);
-  return res;
+InstallGlobalFunction("EvalString",
+function(string)
+    local instream, instring, res;
+    instring := Concatenation(string, ";");
+    instream := InputTextString(instring);
+
+    res := READ_ALL_COMMANDS(instream, false);
```

Comment:
An alternative implementation, which would not break compatibility, could perhaps be written by using `ReadAsFunction`: just prepend `return ` to the string to be evaluated; then use `ReadAsFunction`; an error there indicates a syntax error resp. a non-expression. Finally, execute the function. Only the first part of the string will be executed then (this would still be a difference compared to EvalString, but IMHO one that is an improvement, and does not contradict the existing documentation)